conda_pypi/main.py

```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+                        wheel.setdefault(key, []).append(value.strip())
+                    else:
+                        wheel[key] = value.strip()
+        if ignore:
+            continue
+        if record.url:
+            lines.append(f"# pypi: {record.url}")
+        else:
+            seen = {"abi": set(), "platform": set()}
+            lines.append(f"# pypi: {record.name}=={record.version}")
+            lines[-1] += f" --python-version {python_details['version']}"
+            lines[-1] += f" --implementation {python_details['implementation']}"
+            if wheel and (wheel_tag := wheel.get("Tag")):
+                for tag in wheel_tag:
+                    _, abi_tag, platform_tag = tag.split("-", 2)
+                    if abi_tag != "none" and abi_tag not in seen["abi"]:
+                        lines[-1] += f" --abi {abi_tag}"
+                        seen["abi"].add(abi_tag)
+                    if platform_tag != "any" and platform_tag not in seen["platform"]:
+                        lines[-1] += f" --platform {platform_tag}"
+                        seen["platform"].add(platform_tag)
+            # Here we could try to run a --dry-run --report some.json to get the resolved URL
+            # but it's not guaranteed we get the exact same source so for now we defer to install
+            # time
+        if checksum and hashed_record:
+            lines[-1] += f" -- --record-checksum={checksum}:{hashed_record}"
+
+    return lines
+
+
+def dry_run_pip_json(args: Iterable[str], force_reinstall: bool = False) -> dict[str, Any]:
+    # pip can output to stdout via `--report -` (dash), but this
+    # creates issues on Windows due to undecodable characters on some
+    # project descriptions (e.g. charset-normalizer, amusingly), which
+    # makes pip crash internally. Probably a bug on their end.
+    # So we use a temporary file instead to work with bytes.
+    json_output = NamedTemporaryFile(suffix=".json", delete=False)
+    json_output.close()  # Prevent access errors on Windows
+
+    cmd = [
+        sys.executable,
+        "-mpip",
+        "install",
+        "--dry-run",
+        "--ignore-installed",
+        *(("--force-reinstall",) if force_reinstall else ()),
+        "--report",
+        json_output.name,
+        "--target",
+        json_output.name + ".dir",
+        *args,
+    ]
+    process = run(cmd, capture_output=True, text=True)
+    if process.returncode != 0:
+        raise CondaError(
+            f"Failed to dry-run pip:\n"
+            f"  command: {shlex.join(cmd)}\n"
+            f"  exit code: {process.returncode}\n"
+            f"  stderr:\n{process.stderr}\n"
+            f"  stdout:\n{process.stdout}"
+        )
+
+    with open(json_output.name, "rb") as f:
+        # We need binary mode because the JSON output might
+        # contain weird unicode stuff (as part of the project
+        # description or README).
+        report = json.loads(f.read())
+    os.unlink(json_output.name)
```

Comment:
this should be done in a big try/finally block to make sure this file is deleted no matter what.

Also the `*.dir` directory created by the --target option above needs cleaning up

Reply:
Done.

> Also the *.dir directory created by the --target option above needs cleaning up

It's never actually created, so it fails to clean up. I didn't add it.